app/src/main/java/com/example/unlibrary/library/LibraryRepository.java

```diff
@@ -343,4 +354,71 @@ public void completeExchange(Request request, Book book, OnSuccessListener<Void>
+                    ArrayList<Request> matchingRequests = new ArrayList<>();
+                    for (Request request : requestsOnBook) {
+                        if (request.getRequester().equals(requestedUID) && !request.getState().equals(Request.State.ARCHIVED)) {
+                            matchingRequests.add(request);
+                            //Sanity Check
+                            if (request.getState().equals(Request.State.BORROWED) || request.getState().equals(Request.State.ACCEPTED )) {
+                                Log.e(TAG, "Unexpected requests were found for book ID " + bookRequestedID + " and requesterID " + requestedUID);
+                                return;
+                            }
+                        }
+                    }
+                    // Check to make sure we actually found the request and there was only one such request
+                    if (matchingRequests.size() == 0) {
+                        Log.e(TAG, "The request was not found for book ID " + bookRequestedID + " and requesterID " + requestedUID);
+                        return;  // We don't show toast here to user. Can we do that (we don't have access to mFailureMsgEvent of ViewModel)?
+                    } else if (matchingRequests.size() != 1) {
+                        Log.e(TAG, "More than one request was returned for book ID " + bookRequestedID + " and requesterID " + requestedUID);
+                        return;  // We don't show toast here to user. Can we do that (we don't have access to mFailureMsgEvent of ViewModel)?
+                    }
+                    Request requestToUpdate = matchingRequests.get(0);
+
+                    // Figure out if there are any other non-archived requests on this book (made by other users)
+                    boolean allOtherRequestsAreArchived = true;
+                    for (Request request : requestsOnBook) {
+                        if (!request.getState().toString().equals(Request.State.ARCHIVED.name()) && !request.getId().equals(requestToUpdate.getId())) {
+                            allOtherRequestsAreArchived = false;
+                            break;
+                        }
+                    }
+
+                    // Get DocumentReferences required for transaction
+                    DocumentReference requestToUpdateDocRef = mDb.collection(REQUESTS_COLLECTION).document(requestToUpdate.getId());
+                    DocumentReference currentBookDocRef = mDb.collection(BOOKS_COLLECTION).document(bookRequestedID);
+                    // Convert to a final variable so it can be referenced later in transaction
+                    final boolean updateBookStatus = allOtherRequestsAreArchived;
+
+                    // Transaction to do both updates at once
+                    mDb.runTransaction((Transaction.Function<Void>) transaction -> {
```

Comment:
Do we need to specify/cast this lambda?

Reply:
What does this mean?

Reply:
```suggestion
                    mDb.runTransaction(transaction -> {
```
can we make do with the above?

Reply:
I tried removing it; it causes an error ans red underlines the onSuccessListener. This is what it says:

Required type: OnSuccessListener <? super Object>
Provided: OnSuccessListener <capture of ? super Void>

Probably best to leave it in?

Reply:
That error seems unrelated to converting it to a lambda

Reply:
I believe it is already a lambda. I think Armi means removing the type-cast before the lambda parameter.

Reply:
Ah i see my bad, hmm regardless I'm not sure why you're getting that error. Ill try to take a look at your branch rn

Reply:
I don't think removing this cast is worth the effort right now.